### ISSUE TITLE
Generating type file for a TypeScript project

### ISSUE THREAD

**Issue by stared** [team-innovation/vue-sfc-rollup#35]
I use it (for multiple components, with TypeScript) and it works like charm.

However, the current build process does not update the type file `./package-name.d.ts`.

Is the file expected to be modified manually, as with the example

`export const BraKetVueSample: VueConstructor<Vue>;`

or is there some script to generate it automatically?

**Comment by mgdodge** [comment]
The type file has to be manually updated at this point. I am unsure at this moment how to get it to be generated automatically, but if you know a way I please submit a pull request.

**Comment by stared** [comment]
I see.

Unfortunately, I don't know how to do it automatically. Babel (used by rollup) can transpile TypeScript, but cannot generate the file (https://github.com/babel/babel/issues/9850, https://github.com/babel/babel/issues/9668).

There is https://www.npmjs.com/package/rollup-plugin-typescript2, but I am not knowledgable enough in Rollup to use it in the current build process.

At the same time, my approaches with using `tsc` directly failed to produce a single file with types.

**Comment by ShaggyTech** [comment]
Hi @stared,

I'm working on an NPM package right now and had similar questions related to yours when trying to get Typescript and Rollup to play nice together.  I wanted to share my project with you in the hopes it might help.  This is my first time using both of those so I may not be doing everything the correct way.  

[https://github.com/ShaggyTech/nhtsa-api-wrapper](https://github.com/ShaggyTech/nhtsa-api-wrapper)

In that project, the [dist/types](https://github.com/ShaggyTech/nhtsa-api-wrapper/tree/master/dist/types) folder contains the types that are automatically generated from project source files and placed into the `dist/types` folder when running `yarn build` from project root.

It will also combine the automatically generated types along with any types you explicitly export or declare in your source code.  For example: [dist/types/api/actions/DecodeVin.d.ts](https://github.com/ShaggyTech/nhtsa-api-wrapper/blob/master/dist/types/api/actions/DecodeVin.d.ts) was automatically generated from [this source code](https://www.shaggytech.com/nhtsa-api-wrapper/api_actions_DecodeVin.ts.html).

The majority of this functionality comes from your `tsconfig.json` file.  See [Here](https://github.com/ShaggyTech/nhtsa-api-wrapper/blob/master/tsconfig.json) (sorry! mine's messy right now).  These options in particular are important:
```javascript
"declaration": true, /* Generates corresponding '.d.ts' file. */,
"rootDirs": [
    "./src"
] /* List of root folders whose combined content represents the structure of the project at runtime. */,
"typeRoots": [
    "./src/types",
     "./src/@types",
     "node_modules/@types"
] /* List of folders to include type definitions from. */,
    "types": [
     "node",
     "jest",
     "node-fetch"
] /* Type declaration files to be included in compilation. */,
```

I'm also using the `rollup-plugin-typescript2` plugin that you mentioned.  

My [rollup.config.ts](https://github.com/ShaggyTech/nhtsa-api-wrapper/blob/master/rollup.config.ts).   

I've just been playing around with the Rollup config and seeing how it affects the package size and imported size costs, so that config will likely change as I make improvements.  And honestly it's completely possible that I'm doing things the wrong way as I mentioned above.

I just thought maybe my struggles can help someone else!  🤕 ✌️

**Comment by wessberg** [comment]
Hey there. I read your discussion and had to chime in. The things you are struggling with are what originally made me build [`rollup-plugin-ts`](https://github.com/Wessberg/rollup-plugin-ts) which combines Babel, Rollup, and TypeScript and bundles and tree-shakes declarations for your library without any struggles. Based on your discussion, I thought I'd share it with you. Best regards

**Comment by stared** [comment]
@wessberg it may be the thing I am looking for. However - how to use it with other tools? (I admit, I am new-ish to JS build systems, and for me, it is far from trivial which tools get along with each other, in which order, etc)

In particular, in https://github.com/Quantum-Game/bra-ket-vue/blob/master/build/rollup.config.js (generated by `vue-sfc-rollup`) - which lines should be changed?

**Comment by SumNeuron** [comment]
Im struggling to get typescript to work getting errors like:

> // Define typescript interfaces for autoinstaller
>    7 | // eslint-disable-next-line @typescript-eslint/no-explicit-any
> >  8 | interface InstallFunction extends PluginFunction<any> {

**Comment by mgdodge** [comment]
@wessberg Trying out your plugin, looks promising. But how can I disable minification entirely? vue-sfc-rollup currently only minifies the browser output version, as in server/esm contexts there's no real benefit to minified files - the majority of times bundlers themselves will take care of that sort of thing.

**Comment by mgdodge** [comment]
I'm a little torn on this one, honestly. For the most part, single components are all going to have the same typescript export anyway. Due to the nature of Vue single file components, .vue files are shimmed and thus [can't export anything except the component itself](https://github.com/vuejs/vue-loader/issues/1281). So in the case of a single component, being able to generate the typescript definition might not be worth it.

Where things might be nice is in the case of a library - in which case a generated type definition file would mean an author would only have to maintain the barrel file that serves to gather the components together into a list of exports. Unfortunately, every time I try to get some form of generated types from this, I end up with something like this:

```ts
import { PluginFunction } from 'vue';
interface InstallFunction extends PluginFunction<any> {
    installed?: boolean;
}
declare const plugin: {
    install: InstallFunction;
};
export { plugin as default };
export { default as Component1 } from "./component1.vue";
export { default as Component2 } from "./component2.vue";
```
Not exactly what we're looking for - the definition file is just sending people along to the vue file itself anyway.

@wessberg your rollup-plugin-ts library seems to get the job done for my very basic tests when working with only one component - so it's able to create automatically the same boilerplate file that is already included by this utility. But it still falls short with libraries of components, which is where the real benefit would lie.

If there were a way to automatically generate something like the following, I would be all for it, but my expertise with typescript is not there yet.

```ts
import Vue, { VueConstructor, PluginFunction } from 'vue';
interface InstallFunction extends PluginFunction<any> {
    installed?: boolean;
}
declare const plugin: {
    install: InstallFunction;
};
declare const Component1: VueConstructor<Vue>;
declare const Component2: VueConstructor<Vue>
export { plugin as default, Component1, Component2 };
```

**Comment by wessberg** [comment]
@mgdodge, if you've run into a bug with `rollup-plugin-ts`, feel free to [submit an issue](https://github.com/wessberg/rollup-plugin-ts/issues/new) :-).

However, please be aware that TypeScript has a strict set of allowed extensions (which can be slightly customized based on the `CompilerOptions` such as `allowJs` and `resolveJsonModule`), but which doesn't allow arbitrary extensions such as `.vue` unless you declare such modules manually with something like:

```typescript
declare module "*.vue" {
    // Some module declarations in here
}
```

This is also why TypeScript knows nothing about what may live inside a `.vue` file unless you explicitly manually declare those modules.

**Comment by diondree** [comment]
I just tried an alternative approach to this issue. I compile the library with typescript (`tsc`) first and output that in a separate directory (I call mine `dist-compiled`) and then use the `entry.js` file in that directory as the entry for the rollup config. 

As a result my build command becomes
```bash
tsc -p . && rollup --config build/rollup.config.js
```
here's my tsconfig for reference
```json
{
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "strict": true,
    "declaration": true,
    "outDir": "dist-compiled",
    "declarationDir": "dist/types",
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "importHelpers": true,
    "moduleResolution": "node",
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "baseUrl": ".",
    "types": ["node", "vue"],

    "paths": {
      "@/*": ["src/*"]
    },
    "lib": ["esnext", "dom", "dom.iterable", "scripthost"]
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "dist"]
}
```
Hope this helps :smile:

**Comment by mgdodge** [comment]
@diondree do you have a sample repo? As far as I am aware, typescript by itself is unable to recognize .vue files, so I don't see how your first step of compiling with typescript directly is possible. I am happy to be proven wrong, but it sounds like these instructions are not for vue libraries, but for plain javascript/typescript libraries.

**Comment by diondree** [comment]
@mgdodge I thought so as well, actually I'm not even sure how or why it works but here's an example repo I just created:

https://github.com/diondree/vue-ts-lib-starter

**Comment by ckemmler** [comment]
I must admit that I am a bit lost in translation here.

I am very basically trying to use this tool to get me started on a library of vue components using typescript. I guess my first question is: is it going to work? Because right now, it looks like I am stuck with something that's just broken. I.e. I get a module with two distinct exports, such as:

```
export { __vue_component__$1 as CoBaseTaskbar, __vue_component__ as CoVueComponentsSample };
```
but as it is, I am unable to export something that I am able to import in my project. I tried a few variants, as suggested in the above comments, e.g.:

```
import { PluginFunction } from 'vue';
interface InstallFunction extends PluginFunction<any> {
  installed?: boolean;
}
declare const plugin: {
  install: InstallFunction;
};
export { plugin as default };
export { default as CoBaseTaskbar } from "@/lib-components/co-base-taskbar.vue";
export { default as CoVueComponentsSample } from "@/lib-components/co-vue-components-sample.vue";
```

but with that I get the following warning in my IDE:

![image](https://user-images.githubusercontent.com/41804/94456849-11d75d80-01b4-11eb-8d82-67825dc1615e.png)

I'm just not sure what to make out of this.

I personnaly don't mind having to write the `*.d.ts` file myself. I would be very grateful if someone could give me some clues on how to do it.

**Comment by exceeder** [comment]
@mgdodge @diondree This amazing tool and the discussion above helped me a lot, thank you! In the end, I got it all running, if you are interested: https://github.com/exceeder/vuetrex 

I think, vue-sfc-rollup templates need a little bit more thought in the area of TypeScript support. I think, the best approach to solve this issue would be to let user choose, whether or not to use tsc with babel typescript plugin. 

Not only are the type files not generated with the `vue-sfc-rollup@4.0.0-beta.1`, which is, indeed, solved by running `tsc` and ensuring tsconfig.json has
```
 "declaration": true /* Generates corresponding '.d.ts' file. */,
 "declarationMap": false /* Generates a sourcemap for each corresponding '.d.ts' file. */,
 "declarationDir": "dist/types",` /* choose your directory for d.ts files */
 /* ensure "include", "exclude" and "paths" sections combine to what you want to expose. */
```

But also, without running `tsc`, `@babel/preset-typescript` sometimes does a poor job of transpiling typescript. It is not a full compiler, so, it e.g. leaves class field definitions without moving them into constructor, which breaks a vue-cli generated project.
References: 
https://www.typescriptlang.org/docs/handbook/babel-with-typescript.html and 
https://github.com/babel/babel/issues?q=is%3Aissue+is%3Aopen++typescript+class+property

By making developer aware these technologies are complementary we would save a lot of people a lot of time. Somewhere, it should be explained that while tsc is slower, it is rich in functionality, whereas babel typescript plugin can be faster alternative for simpler code when e.g. type definitions can be generated by hand, but it shines for producing variety of compatible builds.
 
So, with tsc it would simply be 
```
 "build": "tsc && cross-env NODE_ENV=production rollup --config build/rollup.config.js",
```
but then, of course, tsc leaves all the .js and map files, which I currently remove semi-manually before creating a release.

**Comment by italodirenzo93** [comment]
I've run up against this issue now in a project of my own. While it's easy enough to just declare your components as `export const MyComponent: VueConstructor`, a wrench really gets thrown into the mix if you're trying to export other things that aren't components (interfaces, types, functions, etc) because they need to be treated differently. Manually writing `*.d.ts` files for these modules is just painful.

I'm looking into bundling all of this other code into a separate module as a workaround. One that exports only Vue components  (`import Plugin, { MyComponent } from 'my-package'`), and one that exports only support code (`import { myCustomFunction, myMixin, myLookupObject } from 'my-package/contrib'`). This way I can run `tsc` on only this subset and generate `*.d.ts` files for them and just declare the components manually.

In any other case, I do believe this issue needs some attention but also have to admit that my knowledge of TS at this level and the magic that makes Vue work with it is a bit lacking.

**Comment by mgdodge** [comment]
I've got a working solution for this longstanding issue, which should be landing in the next release. Just updating the docs a little, and getting another set of eyes on the code. Ended up with something based on `rollup-plugin-typescript2`. Thanks for the patience on this one!

**Comment by mgdodge** [comment]
Just released v4.0.1, which includes auto-generated typescript definitions. Enjoy!
